### PR TITLE
[Date-Time] Multi-day Day view

### PR DESCRIPTION
**Pull request by lorejoh12**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ yarn change`

#### Description of changes
Implementation of multiple days with dateRangeType === DateRangeType.Day. This is for something like a 4 day view- it's not really a week based view, so work week days won't work for it, because starting on a Saturday should wrap the remaining days around to the next week.

We explicitly don't try to fix the case where the dates wrap around to the next month and go off the edge of the grid- we don't want to grow the grid unnecessarily large, and showing potentially 7 weeks is bad, so we just don't show the days that go off the edge.

Previously hover states were always static, but now the classes have to update dynamically as the mouse moves over days

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/10630)

### CODE REVIEW

**Comment by lorejoh12** [comment]
Draft PR while we settle on the design and determine what other changes we might need, I'm just using this to generate the deploy links for people to see

File: packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx [hunk 1]
```diff
@@ -104,9 +105,10 @@ export class CalendarDayGridBase extends BaseComponent<ICalendarDayGridProps, IC
       labelledBy,
       lightenDaysOutsideNavigatedMonth,
       animationDirection
+      // daysToSelectInDayView
```
[the remark this answers is not in the document]
Comment (lorejoh12) [reply packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx]: good catch, removing

File: packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx [hunk 2]
```diff
@@ -115,9 +117,13 @@ export class CalendarDayGridBase extends BaseComponent<ICalendarDayGridProps, IC
       animationDirection: animationDirection,
       animateBackwards: animateBackwards
     });
+    const classNames = this.classNames;
 
     // When the month is highlighted get the corner dates so that styles can be added to them
     const weekCorners: IWeekCorners = this._getWeekCornerStyles(classNames, weeks!);
+    // if (!(dateRangeType === DateRangeType.Day && daysToSelectInDayView && daysToSelectInDayView > 1)) {
+    //   weekCorners = this._getWeekCornerStyles(classNames, weeks!);
+    // }
```
[the remark this answers is not in the document]
Comment (lorejoh12) [reply packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx]: removing

File: packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx [hunk 3]
```diff
@@ -564,14 +608,53 @@ export class CalendarDayGridBase extends BaseComponent<ICalendarDayGridProps, IC
...
+    return style;
+  }
+
+  private isInSameHoverRange = (date1: Date, date2: Date, date1Selected: boolean, date2Selected: boolean): boolean => {
     const { dateRangeType, firstDayOfWeek, workWeekDays } = this.props;
 
     // The hover state looks weird with non-contiguous days in work week view. In work week, show week hover state
     const dateRangeHoverType = dateRangeType === DateRangeType.WorkWeek ? DateRangeType.Week : dateRangeType;
 
+    // we do not pass daysToSelectInDayView because we handle setting those styles dyanamically in onMouseOver
```
[the remark this answers is not in the document]
Comment (lorejoh12) [reply packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx]: this is specifically in the function for "isInSameHoverRange", so only used for hover. The initial selection is still handled in getWeekCornerStyles

**Comment by micahgodbolt** [comment]
@lorejoh12 is this still a draft PR, or is it ready to go one we get @evlevy signoff?

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts [hunk 4]
```diff
@@ -150,9 +150,17 @@ export function compareDatePart(date1: Date, date2: Date): Number {
...
+ * for multiday view. Defaults to 1
  * @returns {Date[]} An array of dates representing the date range containing the specified date.
  */
-export function getDateRangeArray(date: Date, dateRangeType: DateRangeType, firstDayOfWeek: DayOfWeek, workWeekDays?: DayOfWeek[]): Date[] {
+export function getDateRangeArray(
+  date: Date,
+  dateRangeType: DateRangeType,
+  firstDayOfWeek: DayOfWeek,
+  workWeekDays?: DayOfWeek[],
+  daysToSelectInDayView: number = 1
```
Comment (evlevy) [inline packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts]: >daysToSelectInDayView [](start = 2, length = 21)

what happens when this value < 1?

File: packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx [hunk 5]
```diff
@@ -588,23 +667,27 @@ export class CalendarDayGridBase extends BaseComponent<ICalendarDayGridProps, IC
    *
    */
 
-  private getRefsInRangeOfDay = (day: IDayInfo) => {
+  private getDayInfosInRangeOfDay = (day: IDayInfo) => {
```
Comment (evlevy) [inline packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx]: nit: add return type `IDayInfo[]`

File: packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx [hunk 6]
```diff
@@ -588,23 +667,27 @@ export class CalendarDayGridBase extends BaseComponent<ICalendarDayGridProps, IC
...
 
     // gets all the day refs for the given dates
     const dayInfosInRange = weeks!.reduce((accumulatedValue: IDayInfo[], currentWeek: IDayInfo[]) => {
       return accumulatedValue.concat(currentWeek.filter((weekDay: IDayInfo) => dateRange.indexOf(weekDay.originalDate.getTime()) !== -1));
     }, []);
 
+    return dayInfosInRange;
+  };
+
+  private getRefsFromDayInfos = (dayInfosInRange: IDayInfo[]) => {
```
Comment (evlevy) [inline packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx]: nit: add return type `(HTMLElement | null)[]`

File: packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx [hunk 7]
```diff
@@ -615,19 +698,38 @@ export class CalendarDayGridBase extends BaseComponent<ICalendarDayGridProps, IC
...
         if (dayRef) {
           dayRef.classList.add('ms-CalendarDay-hoverStyle');
+          if (
+            !dayInfos[index].isSelected &&
+            this.props.dateRangeType === DateRangeType.Day &&
+            this.props.daysToSelectInDayView &&
+            this.props.daysToSelectInDayView > 1
+          ) {
+            // remove the static classes first to overwrite them
+            dayRef.classList.remove(this.classNames.bottomLeftCornerDate!);
```
Comment (evlevy) [inline packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx]: option: you can pass all the class names in a single remove() call:

File: packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx [hunk 8]
```diff
@@ -615,19 +698,38 @@ export class CalendarDayGridBase extends BaseComponent<ICalendarDayGridProps, IC
...
+            dayRef.classList.remove(this.classNames.bottomRightCornerDate!);
+            dayRef.classList.remove(this.classNames.topLeftCornerDate!);
+            dayRef.classList.remove(this.classNames.topRightCornerDate!);
+
+            const classNames = this._calculateRoundedStyles(this.classNames, false, false, index > 0, index < dayRefs.length - 1).trim();
+            if (classNames) {
+              classNames.split(' ').forEach((c: string) => dayRef.classList.add(c));
```
Comment (evlevy) [inline packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx]: option: `dayRef.classList.add(...classNames.split(' '));`

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts (hunk 4 above)
Comment (evlevy) [reply packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts]: > what happens when daysToSelectInDayView < 1?

I tried it... it's bad. Definately should be `Math.max(daysToSelectInDayView, 1)`

File: packages/date-time/src/components/pages/CalendarPage.tsx [hunk 9]
```diff
@@ -115,13 +102,24 @@ export class CalendarPage extends React.Component<{}, {}> {
...
+            <ExampleCard
+              title="Calendar with multiday view using dateRangeType === DateRangeType.Day and daysToSelectInDayView = 4"
+              code={CalendarInlineExampleCode}
+            >
+              <CalendarInlineExample
+                dateRangeType={DateRangeType.Day}
+                highlightCurrentMonth={false}
+                highlightSelectedMonth={true}
+                showGoToToday={true}
+                calendarDayProps={{ daysToSelectInDayView: 4 }}
```
Comment (andrewhwng) [inline packages/date-time/src/components/pages/CalendarPage.tsx]: nit: could we add a quick selector to allow users to change the number of days to select on the demo page?

File: packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx (hunk 7 above)
Comment (lorejoh12) [reply packages/date-time/src/components/CalendarDayGrid/CalendarDayGrid.base.tsx]: Wow I totally missed that. I saw that you couldn't pass them as a single string so I assumed you had to call individually but that makes more sense.

**Pull request merged**